core/coreapi/unixfs.go

```diff
@@ -284,35 +297,58 @@ func (api *UnixfsAPI) processLink(ctx context.Context, linkres ft.LinkResult, se
 		}
 	}
 
-	return lnk
+	return lnk, nil
 }
 
-func (api *UnixfsAPI) lsFromLinksAsync(ctx context.Context, dir uio.Directory, settings *options.UnixfsLsSettings) (<-chan coreiface.DirEntry, error) {
+func (api *UnixfsAPI) lsFromLinksAsync(ctx context.Context, dir uio.Directory, settings *options.UnixfsLsSettings) (<-chan coreiface.DirEntry, <-chan error) {
 	out := make(chan coreiface.DirEntry, uio.DefaultShardWidth)
+	errOut := make(chan error, 1)
 
 	go func() {
 		defer close(out)
+		defer close(errOut)
+
 		for l := range dir.EnumLinksAsync(ctx) {
+			dirEnt, err := api.processLink(ctx, l, settings) // TODO: perf: processing can be done in background and in parallel
+			if err != nil {
+				errOut <- err
+				return
+			}
 			select {
-			case out <- api.processLink(ctx, l, settings): // TODO: perf: processing can be done in background and in parallel
+			case out <- dirEnt:
 			case <-ctx.Done():
 				return
 			}
 		}
 	}()
 
-	return out, nil
+	return out, errOut
 }
 
-func (api *UnixfsAPI) lsFromLinks(ctx context.Context, ndlinks []*ipld.Link, settings *options.UnixfsLsSettings) (<-chan coreiface.DirEntry, error) {
-	links := make(chan coreiface.DirEntry, len(ndlinks))
```

Comment:
Fixed this: Using a channel instead of a slice, just to confirm to an interface, seems like a bad idea.